slider/curve.py

```diff
@@ -0,0 +1,230 @@
+from abc import ABCMeta, abstractmethod
+import math
+import bisect
+
+import numpy as np
+from scipy.misc import comb
+from itertools import accumulate
+
+from .position import Position
+from .utils import lazyval
+
+class Curve(metaclass=ABCMeta):
+    _kind_dispatch = {}
+
+    @classmethod
+    def from_kind_and_points(cls, kind, points):
+        if kind == 'B' and len(set(points)) < len(points):
+            # has red points, should be a MetaCurve not a Bezier
+            kind = 'M'
+        try:
+            subcls = cls._kind_dispatch[kind]
+        except KeyError:
+            raise ValueError(f'unknown curve type: {kind!r}')
+
+        return subcls(points)
+
+    @abstractmethod
+    def __call__(self, t):
+        """Compute the position of the curve at time ``t``.
+
+        Parameters
+        ----------
+        t : float
+            The time along the distance of the curve in the range [0, 1]
+
+        Returns
+        -------
+        position : Position
+            The position of the curve.
+        """
+        raise NotImplementedError('__call__')
+
+    def __init_subclass__(cls):
+        cls._kind_dispatch
+        for meth in super(cls, cls).__abstractmethods__:
+            impl = getattr(cls, meth)
+            if impl.__doc__ is None:
+                impl.__doc__ = getattr(super(cls, cls), meth).__doc__
+
+        for kind in cls.kinds:
+            cls._kind_dispatch[kind] = cls
+
+
+class Bezier(Curve):
+    kinds = 'LB'
+
+    def __init__(self, points):
+        self.points = points
+        self._coordinates = np.array(points).T
+
+    def __call__(self, t):
+        points = self.points
+
+        n = len(points) - 1
+        ixs = np.arange(n + 1)
+        x, y = np.sum(
+            comb(n, ixs) *
+            (1 - t) ** (n - ixs) *
+            t ** ixs *
+            self._coordinates,
+            axis=1,
+        )
+        return Position(x, y)
+
+    @lazyval
+    def length(self):
+        """Approximates length as piecewise linear"""
+        total = 0
+        # todo: choose number of points to reduce error below a bound
+        points = [self(t) for t in np.linspace(0, 1, num=5)]
+        for i in range(len(points) - 1):
+            dx = points[i + 1].x - points[i].x
+            dy = points[i + 1].y - points[i].y
+            total += math.sqrt(dx ** 2 + dy ** 2)
+        return total
+
+
+class MetaCurve(Curve):
```

Comment:
That does work but it means single curve curves will have their length calculated for no reason which is incredibly costly (because bezier curves are a pain in the ass too calculate the length of). I could add a check that doesn't do the length calcs for single curves but then it feels like I may as well have a separate class. Also what will the current Bezier class be called? It does represent a single bezier curve so not calling it Bezier seems a little strange.